File: wlrn.lua
--
require 'torch'

require 'cunn'
require 'optim'

torch.setdefaulttensortype('torch.FloatTensor')

----------------------------------------------------------------------------------------------------
--------------------- parse command line options ---------------------------------------------------
----------------------------------------------------------------------------------------------------

cmd = torch.CmdLine()
cmd:text()
cmd:text("Arguments")
cmd:argument("-m", "path to the model (in Torch7 nn format)")
cmd:argument("-t", "training/validation data-loading routines")
cmd:text("Options")
cmd:option("-w", "", "model write destination (in Torch7 nn format)")
cmd:option("-g", "", "GPU ID")
cmd:option("-n", "128", "number of training rounds")
cmd:option("-l", "0.0001", "learning rate")
cmd:option("-b", "32", "batch size")

params = cmd:parse(arg)

-- move computations to a specific GPU (if requested)
if params.g ~= "" then
	--
	cutorch.setDevice( tonumber(params.g) )
end

----------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------

--
T = torch.load(params.m):cuda()
print('* model architecture:')
print(T)

-- if you don't have cuDNN installed, please comment out the following lines
-- (beware: cunn is *significantly* slower than cuDNN)
usecudnn = true
if usecudnn then
	require 'cudnn'
	cudnn.benchmark = true
	cudnn.fastest = true
	T = cudnn.convert(T, cudnn, function(module)
		return torch.type(module):find('SpatialBatchNormalization') -- don't use cudnn for batch normalization
	end)
end

--
pT, gT = T:getParameters()
print('* the model has ' .. pT:size(1) .. ' parameters')

function model_forward(triplet)
	--
	--
	local na=triplet[1]:size(1)
	local np=triplet[2]:size(1)
	local nn=triplet[3]:size(1)
	--
	local descs = T:forward( torch.cat(triplet, 1) )
	--
	return {
		descs:narrow(1, 1, na),
		descs:narrow(1, na+1, np),
		descs:narrow(1, na+np+1, nn)
	}
end

function model_backward(triplet, dloss)
	--
	--
	return T:backward(torch.cat(triplet, 1), torch.cat(dloss, 1))
end

function select_hard_negatives(triplet)
	--
	--
	local negs = {}
	for i=1, #triplet[3] do
		--
		table.insert(negs, T:forward(triplet[3][i]:cuda()):clone())
	end
	negs = torch.cat(negs, 1)
	--
	local _, inds = torch.max(torch.mm(T:forward(triplet[1]:cuda()), negs:t()), 2)
	inds = inds:squeeze():long()
	--
	return {triplet[1], triplet[2], torch.cat(triplet[3], 1):index(1, inds)}
end

----------------------------------------------------------------------------------------------------
----------------------------- loss computation -----------------------------------------------------
----------------------------------------------------------------------------------------------------

--
thr = 0.8
beta = -math.log(1.0/0.99 - 1)/(1.0-thr)

print('* matching threshold set to ' .. thr)

--
M = nn.Sequential()
M:add( nn.MM(false, true) )
M:add( nn.AddConstant(1.0)):add(nn.MulConstant(0.5) ) -- rescale similarities to [0, 1]
M:add( nn.Sequential():add(nn.AddConstant(-thr)):add(nn.MulConstant(beta)):add(nn.Sigmoid()) ) -- kill all scores below the threshold
M:add( nn.Max(2) )
M:add( nn.Contiguous() )
M:add( nn.Sum() )
-- Instead of adding a small constant, eps, just to the score of the positive bag,
-- we add a 1.0 to both the scores of the positive and negative bags.
-- This prevents the division-by-zero error and can be seen as a form of additive regularization:
-- https://en.wikipedia.org/wiki/Additive_smoothing
M:add( nn.AddConstant(1) )

--
C = nn.ConcatTable()
C:add(nn.Sequential():add(nn.ConcatTable():add(nn.SelectTable(1)):add(nn.SelectTable(3))):add(M:clone('weight','bias', 'gradWeight','gradBias')))
C:add(nn.Sequential():add(nn.ConcatTable():add(nn.SelectTable(1)):add(nn.SelectTable(2))):add(M:clone('weight','bias', 'gradWeight','gradBias')))

--
L = nn.Sequential()
L:add(C)
L:add(nn.CDivTable())
L = L:cuda()

--
function loss_forward(triplet)
	--
	return L:forward(triplet)[1]
end

function loss_backward(triplet)
	--
	return L:backward(triplet, torch.ones(1):cuda())
end

function compute_average_loss(triplets)
	--
	-- switch to validation mode
	--
	T:evaluate()
	L:evaluate()

	--
	local avgloss = 0.0

	for i=1, #triplets do
		--
		local triplet = triplets[i]
		if type(triplet[3]) == 'table' then
			--
			triplet = select_hard_negatives(triplet)
		end
		triplet = {triplet[1]:cuda(), triplet[2]:cuda(), triplet[3]:cuda()}

		--
		local descs = model_forward(triplet)

		avgloss = avgloss + loss_forward(descs)
	end

	avgloss = avgloss/#triplets

	--
	T:clearState()
	L:clearState()

	--
	return avgloss
end

----------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------

function apply_optim_sgd_step(triplets, eta)
	--
	local feval = function(x)
		--
		if x ~= pT then
			pT:copy(x)
		end

		--
		gT:zero()

		local loss = 0.0

		for i=1, #triplets do
			--
			local triplet = {
				triplets[i][1]:cuda(),
				triplets[i][2]:cuda(),
				triplets[i][3]:cuda()
			}

			-- forward pass
			--
			local v = model_forward(triplet)
			--
			loss = loss + loss_forward(v)

			-- backward pass
			--
			local dloss = loss_backward(v)
			--
			model_backward(triplet, dloss)
		end

		--
		loss = loss/#triplets
		gT:div(#triplets)

		-- regularization
		--coeff = 1e-4
		--loss = loss + coeff*torch.norm(x)^2/2
		--gT:add( x:clone():mul(coeff) )

		--
		return loss, gT
	end

	--
	cfg = cfg or {}
	cfg.learningRate = eta
	optim.rmsprop(feval, pT, cfg)
end

function train_with_sgd(triplets, niters, bsize, eta)
	--
	-- switch to train mode
	--
	T:training()
	L:training()

	--
	T:zeroGradParameters()

	--
	for i=1, niters do
		--
		local batch = {}

		for j=1, bsize do
			--
			local triplet = triplets[math.random(1, #triplets)]
			if type(triplet[3]) == 'table' then
				--
				triplet = select_hard_negatives(triplet)
			end
			table.insert(batch, triplet)
		end

		--
		apply_optim_sgd_step(batch, eta)
	end

	--
	T:clearState()
	L:clearState()
end

----------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------

--
get_trn_triplets, get_vld_triplets = dofile(params.t)

--
time = sys.clock()
vtriplets = get_vld_triplets()
time = sys.clock() - time
print('* ' .. #vtriplets .. ' validation triplets generated in ' .. time .. ' [s]')

time = sys.clock()
ebest = compute_average_loss(vtriplets)
elast = ebest
print('* initial validation loss: ' .. ebest)

time = sys.clock() - time
print("    ** elapsed time: " .. time .. " [s]")

--
eta = tonumber(params.l)
bsize = tonumber(params.b)
nrounds = tonumber(params.n)


for i = 1, nrounds do
	--
	--
	print("* ROUND (" .. i .. ")")

	--
	time = sys.clock()
	ttriplets = get_trn_triplets()
	time = sys.clock() - time
	print('    ** ' .. #ttriplets .. ' triplets generated in ' .. time .. ' [s]')

	--
	print('    ** eta=' .. eta .. ', bsize=' .. bsize)

	--
	time = sys.clock()
	train_with_sgd(ttriplets, 512, bsize, eta)
	time = sys.clock() - time

	print("    ** elapsed time: " .. time .. " [s]")

	e = compute_average_loss(ttriplets)
	print("    ** average loss (trn): " .. e)
	e = compute_average_loss(vtriplets)
	print("    ** average loss (vld): " .. e)

	if e<ebest then
		--
		if params.w ~= "" then
			--
			local clone = T:clone()
			clone:float()
			if usecudnn then
				clone = cudnn.convert(clone, nn)
			end
			--
			print("* saving the model to `" .. params.w .. "`")
			torch.save(params.w, clone)
		end

		--
		ebest = e
	end

	--
	elast = e

	--
	ttriplets = {}
	collectgarbage()
end

File: wlrn.py
#
import torch
import torch.nn.functional as F

import sys
import math
import time
import random

#
# parse command line options
#

import argparse
parser = argparse.ArgumentParser()

parser.add_argument('modeldef', type=str, help='a script that define the descriptor-extractor model')
parser.add_argument('tripletgen', type=str, help='a script that generates training and validation triplets')
parser.add_argument('--writepath', type=str, default=None, help='where to write the learned model weights')
parser.add_argument('--loadpath', type=str, default=None, help='path from which to load pretrained weights')
parser.add_argument('--learnrate', type=float, default=1e-4, help='RMSprop learning rate')
parser.add_argument('--batchsize', type=int, default=32, help='batch size')
parser.add_argument('--dataparallel', action='store_true', default=False, help='wrap the model into a torch.nn.DataParallel module for multi-gpu learning')

args = parser.parse_args()

#
# model
#

exec(open(args.modeldef).read())
MODEL = init()
if args.loadpath:
	print('* loading pretrained weights from ' + args.loadpath)
	MODEL.load_state_dict(torch.load(args.loadpath))
if args.dataparallel:
	print('* using nn.DataParallel')
	MODEL = torch.nn.DataParallel(MODEL)
MODEL.cuda()

def model_forward(triplet):
	#
	return [
		MODEL.forward(triplet[0]),
		MODEL.forward(triplet[1]),
		MODEL.forward(triplet[2])
	]

def select_hard_negatives(triplet):
	#
	negs = []
	for t in triplet[2]:
		#
		negs.append(MODEL.forward(torch.autograd.Variable(t.float().cuda(), volatile=True)))
	negs = torch.cat(negs, 0)
	#
	_, inds = torch.max(torch.mm(MODEL.forward(torch.autograd.Variable(triplet[0].float().cuda())), negs.t()), 1)
	inds = inds.data.long().cpu()
	#
	return [triplet[0], triplet[1], torch.cat(triplet[2], 0).index_select(0, inds)]

#
# loss computation
#

thr = 0.8
beta = -math.log(1.0/0.99 - 1)/(1.0-thr)

def loss_forward(triplet):
	# compute similarities and rescale them to [0, 1]
	AP = torch.mm(triplet[0], triplet[1].t()).add(1).mul(0.5)
	AN = torch.mm(triplet[0], triplet[2].t()).add(1).mul(0.5)
	# kill all scores below `thr`
	AP = F.sigmoid(AP.add(-thr).mul(beta))
	AN = F.sigmoid(AN.add(-thr).mul(beta))
	# compute the loss
	return (1 + torch.sum(torch.max(AN, 1)[0]))/(1 + torch.sum(torch.max(AP, 1)[0]))

def compute_average_loss(triplets):
	# switch to evaluation mode
	MODEL.eval()

	#
	totalloss = 0.0

	for triplet in triplets:
		#
		if isinstance(triplet[2], list):
			triplet = select_hard_negatives(triplet)
		triplet = [
			torch.autograd.Variable(triplet[0].float().cuda(), volatile=True),
			torch.autograd.Variable(triplet[1].float().cuda(), volatile=True),
			torch.autograd.Variable(triplet[2].float().cuda(), volatile=True)
		]

		#
		descs = model_forward(triplet)
		loss = loss_forward(descs)

		totalloss = totalloss + loss.data[0]

	#
	return totalloss/len(triplets)

#
#
#

optimizer = torch.optim.RMSprop(MODEL.parameters(), lr=args.learnrate)
batchsize = args.batchsize

def train_with_sgd(triplets, niters):
	# switch to train mode
	MODEL.train()

	#
	for i in range(0, niters):
		#
		optimizer.zero_grad()
		for j in range(0, batchsize):
			#
			triplet = triplets[ random.randint(0, len(triplets)-1) ]
			if isinstance(triplet[2], list):
				triplet = select_hard_negatives(triplet)
			triplet = [
				torch.autograd.Variable(triplet[0].float().cuda()),
				torch.autograd.Variable(triplet[1].float().cuda()),
				torch.autograd.Variable(triplet[2].float().cuda())
			]
			#
			descs = model_forward(triplet)
			loss = loss_forward(descs)
			loss.div(batchsize)
			loss.backward()
		#
		optimizer.step()

#
# initialize stuff
#

print('* tripletgen: ' + args.tripletgen)
exec(open(args.tripletgen).read())
get_trn_triplets, get_vld_triplets = init()

#
# 
#

t = time.time()
vtriplets = get_vld_triplets()
t = time.time() - t
print("* " + str(len(vtriplets)) + " validation triplets generated in " + str(t) + " [s]")

t = time.time()
ebest = compute_average_loss(vtriplets)
print("* initial validation loss: " + str(ebest))

t = time.time() - t
print("    ** elapsed time: " + str(t) + " [s]")

#
nrounds = 128
for i in range(0, nrounds):
	#
	print("* ROUND (" + str(1+i) + ")")

	#
	t = time.time()
	ttriplets = get_trn_triplets()
	t = time.time() - t
	print("    ** " + str(len(ttriplets)) + " triplets generated in " + str(t) + " [s]")

	#
	t = time.time()
	train_with_sgd(ttriplets, 512)
	t = time.time() - t

	print("    ** elapsed time: " + str(t) + " [s]")

	e = compute_average_loss(ttriplets)
	print("    ** average loss (trn): " + str(e))
	e = compute_average_loss(vtriplets)
	print("    ** average loss (vld): " + str(e))

	if e<ebest and args.writepath:
		#
		print("* saving model parameters to `" + args.writepath + "`")
		MODEL.cpu()
		if args.dataparallel:
			torch.save(MODEL.module.state_dict(), args.writepath)
		else:
			torch.save(MODEL.state_dict(), args.writepath)
		MODEL.cuda()
		#
		ebest = e

	#
	ttriplets = []